balancer_conn_wrappers.go

```diff
@@ -20,136 +20,171 @@ package grpc
+	"google.golang.org/grpc/internal/balancer/gracefulswitch"
 	"google.golang.org/grpc/internal/buffer"
 	"google.golang.org/grpc/internal/channelz"
 	"google.golang.org/grpc/internal/grpcsync"
 	"google.golang.org/grpc/resolver"
 )
 
-// scStateUpdate contains the subConn and the new state it changed to.
-type scStateUpdate struct {
-	sc    balancer.SubConn
-	state connectivity.State
-	err   error
-}
-
-// exitIdle contains no data and is just a signal sent on the updateCh in
-// ccBalancerWrapper to instruct the balancer to exit idle.
-type exitIdle struct{}
-
-// ccBalancerWrapper is a wrapper on top of cc for balancers.
+// ccBalancerWrapper is a wrapper on top of cc for balancers. It ensures that
+// method invocations on the underlying balancer happen synchronously and in the
+// same order in which they were received from grpc.
+//
+// It uses the gracefulswitch.Balancer internally to ensure that balancer
+// switches happen in a graceful manner.
+//
 // It implements balancer.ClientConn interface.
 type ccBalancerWrapper struct {
-	cc          *ClientConn
-	balancerMu  sync.Mutex // synchronizes calls to the balancer
-	balancer    balancer.Balancer
-	hasExitIdle bool
-	updateCh    *buffer.Unbounded
-	closed      *grpcsync.Event
-	done        *grpcsync.Event
+	cc *ClientConn
+
+	// balancerMu protects access to the following fields. Any calls on the
+	// underlying balancer must be made with the mutex held. This ensures that we
+	// never call the underlying balancer methods concurrently.
+	balancerMu      sync.Mutex
+	balancer        *gracefulswitch.Balancer
+	curBalancerName string
+
+	updateCh *buffer.Unbounded // Updates written on this channel are processed by watcher().
+	resultCh *buffer.Unbounded // Results of calls to UpdateClientConnState() are pushed here.
+	closed   *grpcsync.Event   // Indicates if close has been called.
+	done     *grpcsync.Event   // Indicates if close has completed its work.
 
 	mu       sync.Mutex
 	subConns map[*acBalancerWrapper]struct{}
 }
 
-func newCCBalancerWrapper(cc *ClientConn, b balancer.Builder, bopts balancer.BuildOptions) *ccBalancerWrapper {
+// newCCBalancerWrapper creates a new balancer wrapper. The underlying balancer
+// is not created until the switchTo() method is invoked.
+func newCCBalancerWrapper(cc *ClientConn, bopts balancer.BuildOptions) *ccBalancerWrapper {
 	ccb := &ccBalancerWrapper{
 		cc:       cc,
 		updateCh: buffer.NewUnbounded(),
+		resultCh: buffer.NewUnbounded(),
 		closed:   grpcsync.NewEvent(),
 		done:     grpcsync.NewEvent(),
 		subConns: make(map[*acBalancerWrapper]struct{}),
 	}
 	go ccb.watcher()
-	ccb.balancer = b.Build(ccb, bopts)
-	_, ccb.hasExitIdle = ccb.balancer.(balancer.ExitIdler)
+	ccb.balancer = gracefulswitch.NewBalancer(ccb, bopts)
 	return ccb
 }
 
-// watcher balancer functions sequentially, so the balancer can be implemented
-// lock-free.
+// updateType indicates the type of update pushed to the watcher goroutine.
+type updateType int
```

Comment:
I think this adds a bit of redundant state.  Why not use the type of the update instead, as before?

Reply:
For future reference, we define empty structs in the control buf for the same purpose: https://github.com/grpc/grpc-go/blob/3cccf6a43b55b4a089e832213b0e34df859e1978/internal/transport/controlbuf.go#L184-L185

IMO this is better as it is safer (avoids blind type assertions that could lead to panics) and has less redundant state.